File: android/app/src/main/java/com/noop/analytics/WorkoutCautionPolicy.kt
package com.noop.analytics

import kotlin.math.abs

/**
 * Sustained-exertion guidance for a user-started workout.
 *
 * Kotlin twin of WorkoutCautionPolicy.swift. HRmax is an exertion reference, not a medical limit, and
 * PAUSE_AND_ASSESS never means a medical event was detected.
 */
class WorkoutCautionPolicy(
    private val config: Config,
    private val startTs: Long,
) {
    data class Config(val hrMax: Double)

    enum class Status { WARMUP, ACTIVE, STALE }
    enum class Level { COMFORTABLE, HIGH, VERY_HIGH }
    enum class Cue { EASE_OFF, PAUSE_AND_ASSESS, RECOVERED }

    data class Output(
        val status: Status,
        val level: Level,
        val smoothedBpm: Double?,
        val sampleArrived: Boolean,
        val cue: Cue?,
    )

    private data class Reading(val ts: Long, val bpm: Int)

    private val buffer = ArrayDeque<Reading>()
    private var lastValidTs: Long? = null
    private var lastAcceptedBpm: Double? = null
    private var pendingJumpBpm: Double? = null
    private var pendingJumpAt: Long? = null
    private var easeSince: Long? = null
    private var pauseSince: Long? = null
    private var lastEaseAt: Long? = null
    private var lastPauseAt: Long? = null
    private var lastCueAt: Long? = null
    private var recoveryArmed = false

    fun update(now: Long, bpm: Int?): Output {
        if (lastValidTs?.let { now - it > STALE_AFTER_SEC } == true) resetAfterGap()

        var sampleArrived = false
        if (bpm != null && accept(bpm, now)) {
            buffer.addLast(Reading(now, bpm))
            lastValidTs = now
            lastAcceptedBpm = bpm.toDouble()
            sampleArrived = true
        }

        while (buffer.firstOrNull()?.ts?.let { it < now - SMOOTHING_WINDOW_SEC } == true) {
            buffer.removeFirst()
        }
        val stale = lastValidTs?.let { now - it > STALE_AFTER_SEC } ?: true
        if (stale || buffer.isEmpty()) {
            return Output(Status.STALE, Level.COMFORTABLE, null, sampleArrived, null)
        }

        val smoothed = median(buffer.map { it.bpm.toDouble() })
        val pauseThreshold = config.hrMax * PAUSE_FRACTION_OF_MAX
        val easeThreshold = config.hrMax * EASE_FRACTION_OF_MAX
        val level = when {
            smoothed >= pauseThreshold -> Level.VERY_HIGH
            smoothed >= easeThreshold -> Level.HIGH
            else -> Level.COMFORTABLE
        }

        if (smoothed >= pauseThreshold) {
            if (pauseSince == null) pauseSince = now
        } else if (smoothed < config.hrMax * PAUSE_RESET_FRACTION_OF_MAX) {
            pauseSince = null
        }
        if (smoothed >= easeThreshold) {
            if (easeSince == null) easeSince = now
        } else if (smoothed < config.hrMax * EASE_RESET_FRACTION_OF_MAX) {
            easeSince = null
        }

        val status = if (now - startTs < WARMUP_SEC) Status.WARMUP else Status.ACTIVE
        if (status != Status.ACTIVE || !sampleArrived) {
            return Output(status, level, smoothed, sampleArrived, null)
        }

        var cue: Cue? = null
        val globalReady = lastCueAt?.let { now - it >= MINIMUM_CUE_GAP_SEC } ?: true
        if (recoveryArmed && smoothed < config.hrMax * EASE_RESET_FRACTION_OF_MAX) {
            cue = Cue.RECOVERED
            recoveryArmed = false
        } else if (
            globalReady &&
            pauseSince?.let { now - it >= PAUSE_DWELL_SEC } == true &&
            (lastPauseAt?.let { now - it >= PAUSE_COOLDOWN_SEC } ?: true)
        ) {
            cue = Cue.PAUSE_AND_ASSESS
            lastPauseAt = now
            lastCueAt = now
            pauseSince = now
            easeSince = now
            recoveryArmed = true
        } else if (
            globalReady &&
            smoothed < pauseThreshold &&
            easeSince?.let { now - it >= EASE_DWELL_SEC } == true &&
            (lastEaseAt?.let { now - it >= EASE_COOLDOWN_SEC } ?: true)
        ) {
            cue = Cue.EASE_OFF
            lastEaseAt = now
            lastCueAt = now
            easeSince = now
            recoveryArmed = true
        }

        return Output(status, level, smoothed, sampleArrived, cue)
    }

    private fun accept(bpm: Int, now: Long): Boolean {
        val value = bpm.toDouble()
        if (
            !config.hrMax.isFinite() ||
            config.hrMax !in 100.0..240.0 ||
            value < MIN_PLAUSIBLE_BPM ||
            value > MAX_PLAUSIBLE_BPM
        ) return false

        val prior = lastAcceptedBpm
        val priorAt = lastValidTs
        if (
            prior == null ||
            priorAt == null ||
            now - priorAt > SMOOTHING_WINDOW_SEC ||
            abs(value - prior) <= MAX_JUMP_BPM
        ) {
            pendingJumpBpm = null
            pendingJumpAt = null
            return true
        }

        val pending = pendingJumpBpm
        val pendingAt = pendingJumpAt
        if (
            pending != null &&
            pendingAt != null &&
            now - pendingAt <= JUMP_CORROBORATION_SEC &&
            abs(value - pending) <= JUMP_CORROBORATION_BPM
        ) {
            pendingJumpBpm = null
            pendingJumpAt = null
            return true
        }
        pendingJumpBpm = value
        pendingJumpAt = now
        return false
    }

    private fun resetAfterGap() {
        buffer.clear()
        easeSince = null
        pauseSince = null
        pendingJumpBpm = null
        pendingJumpAt = null
        recoveryArmed = false
    }

    private fun median(values: List<Double>): Double {
        val sorted = values.sorted()
        val mid = sorted.size / 2
        return if (sorted.size % 2 == 0) {
            (sorted[mid - 1] + sorted[mid]) / 2.0
        } else {
            sorted[mid]
        }
    }

    companion object {
        const val EASE_FRACTION_OF_MAX = 0.90
        const val PAUSE_FRACTION_OF_MAX = 0.97
        const val EASE_RESET_FRACTION_OF_MAX = 0.87
        const val PAUSE_RESET_FRACTION_OF_MAX = 0.94
        const val SMOOTHING_WINDOW_SEC = 12L
        const val STALE_AFTER_SEC = 8L
        const val WARMUP_SEC = 60L
        const val EASE_DWELL_SEC = 45L
        const val PAUSE_DWELL_SEC = 45L
        const val EASE_COOLDOWN_SEC = 5L * 60L
        const val PAUSE_COOLDOWN_SEC = 10L * 60L
        const val MINIMUM_CUE_GAP_SEC = 90L
        const val MIN_PLAUSIBLE_BPM = 25.0
        const val MAX_PLAUSIBLE_BPM = 240.0
        const val MAX_JUMP_BPM = 45.0
        const val JUMP_CORROBORATION_SEC = 5L
        const val JUMP_CORROBORATION_BPM = 8.0
    }
}

File: Packages/StrandAnalytics/Sources/StrandAnalytics/WorkoutCautionPolicy.swift
import Foundation

/// Sustained-exertion guidance for a user-started workout.
///
/// HRmax is a personalized exertion reference, not a medical limit. The policy rejects impossible and
/// uncorroborated jump samples, smooths accepted readings, requires dwell, and rate-limits cues. Its
/// strongest output means "pause and assess how you feel", never "a medical event was detected".
public struct WorkoutCautionPolicy {
    public static let easeFractionOfMax = 0.90
    public static let pauseFractionOfMax = 0.97
    public static let easeResetFractionOfMax = 0.87
    public static let pauseResetFractionOfMax = 0.94
    public static let smoothingWindowSec = 12
    public static let staleAfterSec = 8
    public static let warmupSec = 60
    public static let easeDwellSec = 45
    public static let pauseDwellSec = 45
    public static let easeCooldownSec = 5 * 60
    public static let pauseCooldownSec = 10 * 60
    public static let minimumCueGapSec = 90
    public static let minPlausibleBpm = 25.0
    public static let maxPlausibleBpm = 240.0
    public static let maxJumpBpm = 45.0
    public static let jumpCorroborationSec = 5
    public static let jumpCorroborationBpm = 8.0

    public struct Config: Equatable, Sendable {
        public let hrMax: Double

        public init(hrMax: Double) {
            self.hrMax = hrMax
        }
    }

    public enum Status: String, Equatable, Sendable, Codable {
        case warmup
        case active
        case stale
    }

    public enum Level: String, Equatable, Sendable, Codable {
        case comfortable
        case high
        case veryHigh
    }

    public enum Cue: String, Equatable, Sendable, Codable {
        case easeOff
        case pauseAndAssess
        case recovered
    }

    public struct Output: Equatable, Sendable {
        public let status: Status
        public let level: Level
        public let smoothedBpm: Double?
        public let sampleArrived: Bool
        public let cue: Cue?

        public init(
            status: Status,
            level: Level,
            smoothedBpm: Double?,
            sampleArrived: Bool,
            cue: Cue?
        ) {
            self.status = status
            self.level = level
            self.smoothedBpm = smoothedBpm
            self.sampleArrived = sampleArrived
            self.cue = cue
        }
    }

    private struct Reading {
        let ts: Int
        let bpm: Int
    }

    private let config: Config
    private let startTs: Int
    private var buffer: [Reading] = []
    private var lastValidTs: Int?
    private var lastAcceptedBpm: Double?
    private var pendingJumpBpm: Double?
    private var pendingJumpAt: Int?
    private var easeSince: Int?
    private var pauseSince: Int?
    private var lastEaseAt: Int?
    private var lastPauseAt: Int?
    private var lastCueAt: Int?
    private var recoveryArmed = false

    public init(config: Config, startTs: Int) {
        self.config = config
        self.startTs = startTs
    }

    public mutating func update(now: Int, bpm: Int?) -> Output {
        if let lastValidTs, now - lastValidTs > Self.staleAfterSec {
            resetAfterGap()
        }

        var sampleArrived = false
        if let bpm, accept(bpm: bpm, now: now) {
            buffer.append(Reading(ts: now, bpm: bpm))
            lastValidTs = now
            lastAcceptedBpm = Double(bpm)
            sampleArrived = true
        }

        buffer.removeAll { $0.ts < now - Self.smoothingWindowSec }
        let stale = lastValidTs.map { now - $0 > Self.staleAfterSec } ?? true
        guard !stale, !buffer.isEmpty else {
            return Output(
                status: .stale,
                level: .comfortable,
                smoothedBpm: nil,
                sampleArrived: sampleArrived,
                cue: nil
            )
        }

        let smoothed = median(buffer.map { Double($0.bpm) })
        let pauseThreshold = config.hrMax * Self.pauseFractionOfMax
        let easeThreshold = config.hrMax * Self.easeFractionOfMax
        let level: Level = smoothed >= pauseThreshold
            ? .veryHigh
            : (smoothed >= easeThreshold ? .high : .comfortable)

        if smoothed >= pauseThreshold {
            if pauseSince == nil { pauseSince = now }
        } else if smoothed < config.hrMax * Self.pauseResetFractionOfMax {
            pauseSince = nil
        }
        if smoothed >= easeThreshold {
            if easeSince == nil { easeSince = now }
        } else if smoothed < config.hrMax * Self.easeResetFractionOfMax {
            easeSince = nil
        }

        let status: Status = now - startTs < Self.warmupSec ? .warmup : .active
        guard status == .active, sampleArrived else {
            return Output(
                status: status,
                level: level,
                smoothedBpm: smoothed,
                sampleArrived: sampleArrived,
                cue: nil
            )
        }

        var cue: Cue?
        let globalReady = lastCueAt.map { now - $0 >= Self.minimumCueGapSec } ?? true
        if recoveryArmed, smoothed < config.hrMax * Self.easeResetFractionOfMax {
            cue = .recovered
            recoveryArmed = false
        } else if globalReady,
                  let since = pauseSince,
                  now - since >= Self.pauseDwellSec,
                  lastPauseAt.map({ now - $0 >= Self.pauseCooldownSec }) ?? true {
            cue = .pauseAndAssess
            lastPauseAt = now
            lastCueAt = now
            pauseSince = now
            easeSince = now
            recoveryArmed = true
        } else if globalReady,
                  smoothed < pauseThreshold,
                  let since = easeSince,
                  now - since >= Self.easeDwellSec,
                  lastEaseAt.map({ now - $0 >= Self.easeCooldownSec }) ?? true {
            cue = .easeOff
            lastEaseAt = now
            lastCueAt = now
            easeSince = now
            recoveryArmed = true
        }

        return Output(
            status: status,
            level: level,
            smoothedBpm: smoothed,
            sampleArrived: sampleArrived,
            cue: cue
        )
    }

    private mutating func accept(bpm: Int, now: Int) -> Bool {
        let value = Double(bpm)
        guard config.hrMax.isFinite,
              (100...240).contains(config.hrMax),
              value >= Self.minPlausibleBpm,
              value <= Self.maxPlausibleBpm else {
            return false
        }

        guard let prior = lastAcceptedBpm,
              let priorAt = lastValidTs,
              now - priorAt <= Self.smoothingWindowSec,
              abs(value - prior) > Self.maxJumpBpm else {
            pendingJumpBpm = nil
            pendingJumpAt = nil
            return true
        }

        if let pending = pendingJumpBpm,
           let pendingAt = pendingJumpAt,
           now - pendingAt <= Self.jumpCorroborationSec,
           abs(value - pending) <= Self.jumpCorroborationBpm {
            pendingJumpBpm = nil
            pendingJumpAt = nil
            return true
        }
        pendingJumpBpm = value
        pendingJumpAt = now
        return false
    }

    private mutating func resetAfterGap() {
        buffer.removeAll(keepingCapacity: true)
        easeSince = nil
        pauseSince = nil
        pendingJumpBpm = nil
        pendingJumpAt = nil
        recoveryArmed = false
    }

    private func median(_ values: [Double]) -> Double {
        let sorted = values.sorted()
        let mid = sorted.count / 2
        return sorted.count.isMultiple(of: 2)
            ? (sorted[mid - 1] + sorted[mid]) / 2
            : sorted[mid]
    }
}
